gavel_ci/templates/console.html

```diff
@@ -0,0 +1,89 @@
+{% extends "layout.html" %}
+{% block body %}
+<h1>Gavel CI Build: <a href="{{url_for('jobserv.project', proj=proj)}}">{{proj}}</a> <a href="{{url_for('jobserv.build',proj=proj, build=build)}}">#{{build}}</a> - {{run}}</h1>
+<h3>
+    console.log
+    <span id='console-status'></span>
+    <span id='console-raw' style="display:none;">
+        (<a 
+          href="{{url_for('jobserv.run_artifact', proj=proj, build=build, run=run, p='console.log')}}"
+          download="{{proj}}-{{build}}-{{run}}-console.log">raw</a>)
+    </span>
+</h3>
+<div>
+    <pre
+        id="console-log-tail"
+        style="white-space: pre-wrap;"
+    ></pre>
+</div>
+{% endblock %}
+{% block script %}
+{{super()}}
+<script type="text/javascript">
+(function () {
+    let tailElement = $("#console-log-tail");
+    let downloadElement = $("#console-raw");
+    let statusElement = $("#console-status");
+    var offset = 0;
+    let messageTailWindow = 40;
+    var messageTail = [];
+    var status = '';
+    var interval = 1000; /* 1s */
+    let url = "/projects/{{proj}}/builds/{{build}}/{{run}}/console/tail";
+    var showStatus = function (slug) {
+        let status = {
+            running: 'tailing',
+            queued: 'starting',
+        }[slug.toLowerCase()] || 'tail';
+        $(statusElement).text(status);
+    };
+    var showDownload = function () {
+        $(downloadElement).show();
+    }
+    var updateLog = function (message) {
+        if (!message) return;
+        const messageRows = message.split('\n');
+        messageTail = messageTail.concat(messageRows);
+        messageTail = messageTail.slice(Math.max(messageTail.length - messageTailWindow, 0));
+        let messageContent = messageTail.join('\n');
+        $(tailElement).text(`...\n${messageContent}`);
```

Comment:
So I guess the idea here is to only show `messageTailWindow` rows of console log? Is the issue that its too "expensive" to keep the entire log displayed?

Reply:
There is the "unknown" of how long a log would be and what would cause display issues for the user.  It also lets the window stay a fixed size and does a scroll behavior similar to a terminal window -- while retaining the rest of the website visible. There is a little loss in the inability to scroll up, but I figure that is minimal since we're making the console.log available a download

Reply:
this seems reasonable. I may have some bikeshedding like "make it a bigger amount of lines". However, in general this may make good sense. Its just "different", so I'll need a moment to play with it and absorb